appshell/appshell_extensions_platform.cpp

```diff
@@ -27,5 +37,52 @@ void GetCharsetMatch(const char* bufferData, size_t bufferLength, std::string &d
 	const char* detectedLanguage = ucsdet_getLanguage(charsetMatch_, &icuError);
 	// Get Confidence
 	int32_t detectionConfidence = ucsdet_getConfidence(charsetMatch_, &icuError);
-	ucsdet_close(charsetDetector_);
-}
\ No newline at end of file
+}
+
+CharSetEncode::CharSetEncode(std::string encoding) {
+    status = U_ZERO_ERROR;
+    conv = ucnv_open(encoding.c_str(), &status);
+    if (U_FAILURE(status)) {
+        throw "Unable to open Converter";
+    }
+}
+
+CharSetEncode::~CharSetEncode() {
+    if (conv) {
+        ucnv_close(conv);
+    }
+}
+
+void CharSetEncode::operator()(std::string &contents) {
+    UnicodeString ustr(contents.c_str());
+    int targetLen = ustr.extract(NULL, 0, conv, status);
+    if(status != U_BUFFER_OVERFLOW_ERROR) {
+        throw "Unable to convert encoding";
+    }
+    std::auto_ptr<char> target(new char[targetLen + 1]());
+    status = U_ZERO_ERROR;
+    ustr.extract(target.get(), targetLen, conv, status);
+    target.get()[targetLen] = '\0';
+    contents.assign(target.get(), targetLen);
+}
+
+#if defined(OS_MACOSX) || defined(OS_LINUX)
+void DecodeContents(std::string &contents, std::string encoding) {
+    UnicodeString ustr(contents.c_str(), encoding.c_str());
+    UErrorCode status = U_ZERO_ERROR;
+    int targetLen = ustr.extract(NULL, 0, NULL, status);
+    if(status != U_BUFFER_OVERFLOW_ERROR) {
+        throw "Unable to decode contents";
+    }
+    std::auto_ptr<char> target(new char[targetLen + 1]());
```

Comment:
you may have to use `std::unique_ptr` here with C++11 support. `auto_ptr` is deprecated